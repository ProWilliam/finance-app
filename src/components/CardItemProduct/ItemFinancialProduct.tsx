import React, { useState } from 'react';
import { TouchableOpacity, Text, View } from 'react-native';



// Hook and Component
import useAppNavigation from '../../hooks/useAppNavigation';
import IconAngleRight from '../../assets/icons/IconAngleRight';

// Styles
import style from './ItemFinancialProduct.styles';
import ItemFinancialProductProps from '../../types/components/ItemFinancialProduct.types';

const ItemFinancialProduct: React.FC<ItemFinancialProductProps> = ({ id, name }) => {

  const [route, setRoute] = useState('');
  const { appNavigation } = useAppNavigation();

  appNavigation(route, {id});

  return (
    <TouchableOpacity 
      style={style.containerItem}
      onPress={() => {setRoute('info')}}   
    >
      <View>
        <Text style={style.text}>{name}</Text>
        <Text style={style.textId}>ID:{id}</Text>
      </View>
      <IconAngleRight style={style.icon} />
    </TouchableOpacity>
  );
}

export default ItemFinancialProduct;