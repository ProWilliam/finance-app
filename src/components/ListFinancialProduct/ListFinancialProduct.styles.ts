import { StyleSheet } from 'react-native';
import theme from '../../theme/theme';

const StyleListFinance = StyleSheet.create({

  containerList : {
    width: `${theme.withScreen.medium}%`,

    borderWidth: 1,
    borderColor: theme.light.border,
    borderRadius: 8,

    marginBottom: 100,
  },
  separator: {
    width: `${theme.withScreen.big}%`,
    height: 2,
    backgroundColor: theme.light.backgroundSecondary 
  } 
});

export default StyleListFinance;