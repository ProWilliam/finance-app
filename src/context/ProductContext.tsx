import React, { createContext, useState, useContext } from 'react';

// Types
import { GlobalProviderProps, ProductContextInterface, Product } from './ProductContext.type';


// Default State Context by products.
const defaultStateProduct = {
  products: [{
    id: '',
    name: ''
  }],
  setProducts: (product: Product[]) => {}

} as ProductContextInterface

const MyContextProducts = createContext<ProductContextInterface>(defaultStateProduct);

export const useMyContextProduct = () => {
  const context = useContext(MyContextProducts);

  if(!context) {
    throw new Error('useMycontext must be used within a MyContextProvider');
  }

  return context;
}

export const ProductProvider: React.FC<GlobalProviderProps> = ({ children }) => {
  const [ products, setProducts ] = useState<Product[]>([]);

  return (
    <MyContextProducts.Provider value={{ products, setProducts }}>
      {children}
    </MyContextProducts.Provider>
  );
};