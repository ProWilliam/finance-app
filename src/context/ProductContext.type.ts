import { ReactNode, Dispatch, SetStateAction } from 'react';

export type Product = {
  id: string,
  name: string
}

export interface ProductContextInterface {
  products: Product[];
  setProducts: Dispatch<SetStateAction<Product[]>>;
}

export interface GlobalProviderProps {
  children: ReactNode;
}