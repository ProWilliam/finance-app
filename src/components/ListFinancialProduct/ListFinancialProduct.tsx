import React from 'react';
import { FlatList, View } from 'react-native';

// Components
import ItemFinancialProduct from '../CardItemProduct/ItemFinancialProduct';

// Styles and Types
import style from './ListFinancialProduct.styles';
import { FinanceProductsProps } from '../../types/components/ListFinancialProduct.types';


const ListFinancialProduct: React.FC<FinanceProductsProps> = ({ data }) => {
  return(
    <FlatList 
      style={style.containerList}
      data={data}
      ItemSeparatorComponent={() => <View style={style.separator}/>}
      renderItem={({ item: product}) => 
        <ItemFinancialProduct id={product.id} name={product.name}/>
      }
    />
  )
}


export default ListFinancialProduct;