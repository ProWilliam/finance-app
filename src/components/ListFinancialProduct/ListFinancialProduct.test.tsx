import React from 'react';
import renderer, { act } from 'react-test-renderer';

// Context and Components
import { useMyContextProduct} from '../../context/ProductContext';
import ListFinancialProduct from './ListFinancialProduct';
import ItemFinancialProduct from '../CardItemProduct/ItemFinancialProduct';

// Types and Styles
import { Product } from '../../context/ProductContext.type';
import styles from './ListFinancialProduct.styles';

jest.mock('../../context/ProductContext', () => ({
  useMyContextProduct: jest.fn(),
}));

// Mock useAppNavigation hook
const MockuseAppNavigation = jest.fn();
jest.mock('../../hooks/useAppNavigation', () => ({
  __esModule: true,
  default: () => ({
    appNavigation: MockuseAppNavigation,
  }),
}));

describe('<ListFinancialProduct />', () => {

  let mockProducts: Product[];

  afterEach(() => {
    mockProducts = [
      { id: '1', name: 'Product 1' },
      { id: '2', name: 'Product 2' },
    ];

    jest.clearAllMocks();
  }); 

  it('renders correctly with products', () => {
    
    (useMyContextProduct as jest.Mock).mockReturnValue({ products: mockProducts });

    let tree;

    act(() => {
      tree = renderer.create(<ListFinancialProduct />).toJSON();
    });
  
    expect(tree).toMatchSnapshot();
  });

  it('does not render anything if products is null or undefined', () => {
    (useMyContextProduct as jest.Mock).mockReturnValue({ products: null });

    const tree =  renderer.create(<ListFinancialProduct />).toJSON();

    expect(tree).toBeNull();
  });

  it('renders the correct number of ItemFinancialProduct components', () => {

    (useMyContextProduct as jest.Mock).mockReturnValue({ products: mockProducts });
    MockuseAppNavigation.mockReturnValue({ appNavigation: mockProducts })  
    
    const tree =  renderer.create(
      <ListFinancialProduct />
    );
    
    const items =  tree.root.findAllByType(ItemFinancialProduct);

    expect(items).toHaveLength(mockProducts.length);
  });

  it('renders the correct separator component', () => {
    (useMyContextProduct as jest.Mock).mockReturnValue({ products: mockProducts });

    const tree =  renderer.create(<ListFinancialProduct />);
    const separators = tree.root.findAllByProps({ style: styles.separator});
    
    expect(separators).toHaveLength(mockProducts.length);
  });
});
