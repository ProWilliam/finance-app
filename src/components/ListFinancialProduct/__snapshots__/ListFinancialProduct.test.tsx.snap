// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListFinancialProduct /> renders correctly with products 1`] = `null`;
